src/core/geometry/qgsquadix.cpp

```diff
@@ -0,0 +1,291 @@
+
+      rect.setPoints( p1, p2, p2.project( distance, azimuth, inclination ), p1.project( distance, azimuth, inclination ) );
+      break;
+    }
+  }
+
+  return rect;
+
+}
+
+QgsQuadix QgsQuadix::rectangleFromExtent( const QgsPoint &p1, const QgsPoint &p2 )
+{
+  double z = p1.z();
+
+  double xMin = 0, xMax = 0, yMin = 0, yMax = 0;
+
+  if ( p1.x() < p2.x() )
+  {
+    xMin = p1.x();
+    xMax = p2.x();
+  }
+  else
+  {
+
+    xMin = p2.x();
+    xMax = p1.x();
+  }
+
+  if ( p1.y() < p2.y() )
+  {
+    yMin = p1.y();
+    yMax = p2.y();
+  }
+  else
+  {
+
+    yMin = p2.y();
+    yMax = p1.y();
+  }
+
+  return QgsQuadix( QgsPoint( xMin, yMin, z ),
+                    QgsPoint( xMin, yMax, z ),
+                    QgsPoint( xMax, yMax, z ),
+                    QgsPoint( xMax, yMin, z ) );
+}
+
+QgsQuadix QgsQuadix::squareFromDiagonal( const QgsPoint &p1, const QgsPoint &p2 )
+{
+  QgsPoint point2, point3 = QgsPoint( p2.x(), p2.y() ), point4;
+
+  double azimuth = p1.azimuth( p2 ) + 90.0;
+  double distance = p1.distance( p2 ) / 2.0;
+  QgsPoint midPoint = QgsGeometryUtils::midpoint( p1, p2 );
+
+  point2 = midPoint.project( -distance, azimuth );
+  point4 = midPoint.project( distance, azimuth );
+
+  if ( p1.is3D() )
+  {
+    double z = 0;
+    z = p1.z();
+    point3.addZValue( z );
+    point2.addZValue( z );
+    point4.addZValue( z );
+  }
+
+  return QgsQuadix( p1, point2, point3, point4 );
+
+}
+
+QgsQuadix QgsQuadix::rectangleFromCenterPoint( const QgsPoint &center, const QgsPoint &point )
+{
+
+  double xOffset = fabs( point.x() - center.x() );
+  double yOffset = fabs( point.y() - center.y() );
+
+  return QgsQuadix( QgsPoint( center.x() - xOffset, center.y() - yOffset, center.z() ),
+                    QgsPoint( center.x() - xOffset, center.y() + yOffset, center.z() ),
+                    QgsPoint( center.x() + xOffset, center.y() + yOffset, center.z() ),
+                    QgsPoint( center.x() + xOffset, center.y() - yOffset, center.z() ) );
+
+}
+
+QgsQuadix QgsQuadix::fromRectangle( const QgsRectangle &rectangle )
+{
+  return QgsQuadix(
+           QgsPoint( rectangle.xMaximum(), rectangle.yMinimum() ),
+           QgsPoint( rectangle.xMinimum(), rectangle.yMaximum() ),
+           QgsPoint( rectangle.xMaximum(), rectangle.yMaximum() ),
+           QgsPoint( rectangle.xMaximum(), rectangle.yMinimum() )
+         );
+}
+
+bool QgsQuadix::operator==( const QgsQuadix &other ) const
+{
+  if ( isEmpty() && other.isEmpty() )
+  {
+    return true;
+  }
+  else if ( isEmpty() || other.isEmpty() )
+  {
+    return false;
+  }
+  return ( ( mPoint1 == other.mPoint1 ) &&
+           ( mPoint2 == other.mPoint2 ) &&
+           ( mPoint3 == other.mPoint3 ) &&
+           ( mPoint4 == other.mPoint4 )
+         );
+}
+
+bool QgsQuadix::operator!=( const QgsQuadix &other ) const
+{
+  return !operator==( other );
+}
+
+bool QgsQuadix::isEmpty() const
+{
+  return ( ( mPoint1 == QgsPoint() ) &&
```

Comment:
This is deceptive -- a QgsPoint() is a QgsPoint( 0, 0 ) -- there's no concept of a "null point" in the API. So this method will say a (0,0) x 4 quadix is empty, but a (1,1) x 4 quadix is not.

Reply:
Yes, I forgot that point. Anyway, it's not over yet. The logic will be changed since the class only accepts simple quadrilaterals (without auto intersection) and having no colinear points.